File: Files/src/cl/ejeldes/io/file/java/FileReading.java
package cl.ejeldes.io.file.java;

import java.io.*;
import java.nio.charset.StandardCharsets;

/**
 * Created by ejeldes on Aug 11, 2019
 */
public class FileReading {

    public static void main(String[] args) {
        try {
            InputStreamReader inputStreamReader = new InputStreamReader(
                    new FileInputStream(new File("testfile.txt")), StandardCharsets.UTF_8);
            int character;
            while ((character = inputStreamReader.read()) != -1) {
                char word = (char) character;
                System.out.print(word);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: Files/src/cl/ejeldes/io/file/kotlin/FileReading.kt
package cl.ejeldes.io.file.kotlin

import java.io.File

/**
 * Created by ejeldes on Aug 11, 2019
 */
fun main() {
    // same in java as
    //InputStreamReader inputStreamReader = new InputStreamReader(
    //                    new FileInputStream(new File("testfile.txt")), StandardCharsets.UTF_8);
    // This function takes all the lines in memory so is not good for huge files
    val lines = File("testfile.txt").reader().readLines()
    lines.forEach { println(it) }

    // readText() returns the entire file as a String but you have to close the reader
    // val reader = File("testfile.txt").reader()
    // val text: String = reader.readText()
    // reader.close()
    // The use() method closes the resource properly
    val text = File("testfile.txt").reader().use { it.readText() }
    println(text)

    val file = File("testfile.txt").bufferedReader().use {
        var line: String
        do {
            line = it.readLine()
        } while (!line.contains("3"))
        println("This is line $line")
    }
}